### PR TITLE
Lido app/marcus/lido 3287

### PR DESCRIPTION
**Pull request by marhong2**
This change breaks COUNT and COUNTIFS (due to the changed input) I will fix them shortly @tyleradams @jmorganmartin

### CODE REVIEW

**Comment by tyleradams** [comment]
Fix them in this PR

File: formulas/functions/date.js [hunk 1]
```diff
@@ -651,9 +651,7 @@ const DateFunctions = {
         );
     }
   },
-  // If something's not a date, just wrap in TO_DATE?
```
Comment (tyleradams) [inline formulas/functions/date.js:654]: Leave this in

File: formulas/operators.js [hunk 2]
```diff
@@ -54,24 +54,28 @@ const type2Number = { boolean: 3, string: 2, number: 1 };
...
     const type1 = typeof value1,
       type2 = typeof value2;
 
     if (type1 === type2) {
       // same type comparison
-      switch (infix) {
-        case "=":
-          return value1 === value2;
+      switch (infix) {          
         case ">":
```
Comment (tyleradams) [inline formulas/operators.js:79]: Let's actually support > and < as well as = in this PR

File: formulas/operators.js [hunk 3]
```diff
@@ -54,24 +54,28 @@ const type2Number = { boolean: 3, string: 2, number: 1 };
 
 const Infix = {
   compareOp: (value1, infix, value2, isArray1, isArray2) => {
+
     if (value1 == null) value1 = 0;
     if (value2 == null) value2 = 0;
+    if(infix === "="){
+      let x = FormulaHelpers.equalOP(value1, value2);
```
Comment (tyleradams) [inline formulas/operators.js:61]: return FormulaHelpers.equalOP(value1, value2);

File: formulas/helpers.js [hunk 4]
```diff
@@ -179,6 +179,46 @@ class FormulaHelpers {
...
+        rv.push(curr)
+      }
+      return rv
+    }
+
+    let array = (Array.isArray(input1)) ? input1 : input2;
+    let constant = (Array.isArray(input1)) ? input2 : input1;
+
+    //for a single value
+    if(array.length === 1 && array[0].length === 1){
```
Comment (tyleradams) [inline formulas/helpers.js:209]: Why do we need to special case this?

Comment (marhong2) [reply formulas/helpers.js:209]: A lot of the current implementations of formulas rely on this comparison. I can change them locally if prefered

File: formulas/operators.js (hunk 3 above)
Comment (marhong2) [reply formulas/operators.js:61]: Yes, sorry this was for testing and I forgot to remove it

File: formulas/functions/statistical.js [hunk 5]
```diff
@@ -33,20 +33,30 @@ const countIf = (range, criteria) => {
...
-        arr[cnt] = true;
+      } else{
+        let x = Infix.compareOp(value, criteria.op, criteria.value, Array.isArray(value), isCriteriaArray)
+        if(typeof x === "boolean"){
+          bool = x
+        }else{
+          let localBool = false
+          
+          //not modifying bool for some reason
+          x.map(function(row) {
```
Comment (tyleradams) [inline formulas/functions/statistical.js:44]: use for each and rocket notation
```
x.forEach(row => 
  row.forEach(val => 
```

File: formulas/functions/statistical.js [hunk 6]
```diff
@@ -33,20 +33,30 @@ const countIf = (range, criteria) => {
...
+            row.map(function(val){
+              localBool = localBool || val;
```
Comment (tyleradams) [inline formulas/functions/statistical.js:46]: localBool |= val

File: formulas/helpers.js [hunk 7]
```diff
@@ -179,6 +179,96 @@ class FormulaHelpers {
...
+      comparator = H.getComparator(comparison)
+      return comparator(input1, input2)
+    }
+    const rv = []
+
+    if(Array.isArray(input1) && Array.isArray(input2)){
+      comparator = H.getComparator(comparison)
+      let maxRow = Math.max(input1.length, input2.length);
+      for(let row = 0; row < maxRow; row++){
+        let row1 = (input1.length > row) ? input1[row] : []
```
Comment (tyleradams) [inline formulas/helpers.js:194]: Use const not let

File: formulas/helpers.js [hunk 8]
```diff
@@ -179,6 +179,96 @@ class FormulaHelpers {
...
+    if(Array.isArray(input1)){
+      array = input1
+      constant = input2
+      comparator = H.getComparator(comparison)
+    }else if(Array.isArray(input2)){
+      array = input2
+      constant = input1
+      comparator = H.getComparator(comparison, true)
+    }else {
+      throw "Impossible to reach error"
```
Comment (tyleradams) [inline formulas/helpers.js:218]: Also throw the values of input1 and input2

File: formulas/helpers.js [hunk 9]
```diff
@@ -179,6 +179,96 @@ class FormulaHelpers {
...
+    }
+
+    for(let row = 0; row < array.length; row++){
```
Comment (tyleradams) [inline formulas/helpers.js:221]: We can simplify this:
```
return array.map(row => row.map(v => comparator(constant, val));
```

File: formulas/helpers.js [hunk 10]
```diff
@@ -179,6 +179,96 @@ class FormulaHelpers {
...
+      }
+      rv.push(curr);
+    }
+    return rv; 
+  }
+  getComparator(comparison, reverse = false){
+    if(reverse){
+      switch(comparison){
+        case "=":
+          return function(a, b) {return a === b}
```
Comment (tyleradams) [inline formulas/helpers.js:234]: Push code into data when possible:
```
const m = {
  "=": (a,b) => a === b,
...
}
```
so the code becomes
```
if(reverse) {
  return (a,b) => map[comparison](b,a);
} else {
  return map[comparison];
}
```

File: formulas/operators.js [hunk 11]
```diff
@@ -56,22 +56,26 @@ const Infix = {
...
     }
-
+      
     const type1 = typeof value1,
       type2 = typeof value2;
 
     if (type1 === type2) {
       // same type comparison
-      switch (infix) {
+      switch (infix) {
```
Comment (tyleradams) [inline formulas/operators.js:76]: Use the same map I suggested above, put it in a util if you have to and import in both locations

**Review by tyleradams — CHANGES_REQUESTED** [review]
Getting closer

**Comment by marhong2** [comment]
@tyleradams @jmorganmartin Changes are implemented :)

File: formulas/helpers.js [hunk 12]
```diff
@@ -179,6 +180,66 @@ class FormulaHelpers {
...
+  getComparator(comparison, reverse = false){
+    const map = {
+        "=": (a, b) => a === b,
+        ">": (a, b) => a > b,
+        "<": (a, b) => a < b,
+        ">=": (a, b) => a >= b,
+        "<=": (a, b) => a <= b,
+        "<>": (a,b) => a !== b,
+    }
+    //order of the comparisons matter 1>2 != 2>1,
```
Comment (tyleradams) [inline formulas/helpers.js:232]: duh, no need for a comment

File: formulas/helpers.js [hunk 13]
```diff
@@ -179,6 +180,66 @@ class FormulaHelpers {
...
+    const map = {
+        "=": (a, b) => a === b,
+        ">": (a, b) => a > b,
+        "<": (a, b) => a < b,
+        ">=": (a, b) => a >= b,
+        "<=": (a, b) => a <= b,
+        "<>": (a,b) => a !== b,
+    }
+    //order of the comparisons matter 1>2 != 2>1, 
+    //This allows us to swap orderings of our parameters
```
Comment (tyleradams) [inline formulas/helpers.js:233]: obvious from code

File: formulas/operators.js [hunk 14]
```diff
@@ -56,52 +56,29 @@ const Infix = {
...
-          return type2Number[type1] < type2Number[type2];
-        case "<>":
-          return true;
-        case "<=":
-          return type2Number[type1] <= type2Number[type2];
-        case ">=":
-          return type2Number[type1] >= type2Number[type2];
-      }
-    }
+    if (type1 === type2)
```
Comment (tyleradams) [inline formulas/operators.js:72]: So short, gj

File: formulas/operators.js [hunk 15]
```diff
@@ -56,52 +56,29 @@ const Infix = {
...
-          return true;
-        case "<=":
-          return type2Number[type1] <= type2Number[type2];
-        case ">=":
-          return type2Number[type1] >= type2Number[type2];
-      }
-    }
+    if (type1 === type2) 
+      return comparator(value1, value2)
+    else
```
Comment (tyleradams) [inline formulas/operators.js:74]: No need for the else

**Review by tyleradams — APPROVED** [review]

**Pull request merged**